lib/amqp/client/connection.rb

```diff
@@ -156,6 +156,19 @@ def parse_frame(type, channel_id, frame_size, buf)
             tag_len = buf.unpack1("@11 C")
             tag = buf.byteslice(12, tag_len).force_encoding("utf-8")
             @channels[channel_id].reply [:basic_cancel_ok, tag]
+          when 50 # return
+            reply_code, reply_text_len = buf.unpack("@11 S> C")
+            pos = 14
+            reply_text = buf.byteslice(pos, reply_text_len).force_encoding("utf-8")
+            pos += reply_text_len
+            exchange_len = buf[pos].ord
+            pos += 1
+            exchange = buf.byteslice(pos, exchange_len).force_encoding("utf-8")
+            pos += exchange_len
+            routing_key_len = buf[pos].ord
+            pos += 1
+            routing_key = buf.byteslice(pos, routing_key_len).force_encoding("utf-8")
+            @channels[channel_id].return ReturnMessage.new(reply_code, reply_text, exchange, routing_key)
```

Comment:
After the Return frame, a Header frame and optionally one or more Body frames will arrive, that belongs to this ReturnMessage, so you have to read them like this: https://github.com/cloudamqp/amqp-client.rb/blob/a5091e93b863f607f3766b8d0812053844c651f1/lib/amqp/client/channel.rb#L241-L247